### ISSUE TITLE
Report: Error during unshare(CLONE_NEWUSER): Operation not permitted in the non-privileged container of k8s

### ISSUE THREAD

**Issue by zhangtbj** [containers/buildah#2262]
Hi all,

I need to run the buildah to build my source code on a shared kube cluster.

There are serval security policise and cannot run the container with privileged.

So when I running the buildah in the non-privileged container of k8s.

It reports the error:
```
...
time="2020-03-31T13:30:14Z" level=warning msg="The cgroups manager is set to systemd but there is no systemd user session available"
time="2020-03-31T13:30:14Z" level=warning msg="For using systemd, you may need to login using an user session"
time="2020-03-31T13:30:14Z" level=warning msg="Alternatively, you can enable lingering with: `loginctl enable-linger 1000` (possibly as root)"
time="2020-03-31T13:30:14Z" level=warning msg="Falling back to --cgroup-manager=cgroupfs"
Error during unshare(CLONE_NEWUSER): Operation not permitted
level=error msg="error parsing PID \"\": strconv.Atoi: parsing \"\": invalid syntax"
level=error msg="(unable to determine exit status)"
2020/03/31 13:30:14 Skipping step because a previous step failed
2020/03/31 13:30:15 Skipping step because a previous step failed
{"level":"info","ts":1585661399.569302,"caller":"creds-init/main.go:44","msg":"Credentials initialized."}
```

I also tried this `--isolation=chroot` parameter for `buildah bud` command, but doesn't work.

And I asked open all allowedCapabilities also doesn't work:
```
spec:
  allowPrivilegeEscalation: true
  allowedCapabilities:
  - '*'
```

I run the buildah by using Tekton task:
https://github.com/tektoncd/catalog/blob/master/buildah/buildah.yaml#L43

I saw there are many people had this problem before, but I didn't see the final solution: https://github.com/containers/buildah/issues/1011

I think buildah should work like other builder like kaniko that can build without privileged.

Can someone help me?

Thanks a lot!

**Comment by rhatdan** [comment]
Do you know if this running with docker backend or cri-o?  Do you know if seccomp.json rules are being applied.
What OS is the server running with?

**Comment by zhangtbj** [comment]
Hi @rhatdan ,

We are running buildah as Tekton task on the Kubernetes which based on the `containerd`

It should be the `ubuntu` OS and I saw we set the seccomp setting in the kube PSP annotations like below.

```
kind: PodSecurityPolicy
metadata:
  annotations:
    apparmor.security.beta.kubernetes.io/allowedProfileNames: runtime/default
    apparmor.security.beta.kubernetes.io/defaultProfileName: runtime/default
    seccomp.security.alpha.kubernetes.io/allowedProfileNames: docker/default
    seccomp.security.alpha.kubernetes.io/defaultProfileName: docker/default
spec:
  allowPrivilegeEscalation: false

  defaultAddCapabilities:
  - CHOWN
  - DAC_OVERRIDE
  - FOWNER

  forbiddenSysctls:
  - '*'
```

But it doesn't work ...., Any suggestion? Or did I miss any other seccomp config here?

Thanks!

**Comment by zhangtbj** [comment]
After I changed the seccomp annotations in kube PSP like this:
```
    seccomp.security.alpha.kubernetes.io/allowedProfileNames: '*'
    seccomp.security.alpha.kubernetes.io/defaultProfileName: unconfined
```

The buildah build change to a new error:
```
Error: error writing "0 0 4294967295\n" to /proc/21/gid_map: write /proc/21/gid_map: operation not permitted
level=error msg="error writing \"0 0 4294967295\\n\" to /proc/21/gid_map: write /proc/21/gid_map: operation not permitted"
level=error msg="(unable to determine exit status)"
2020/04/01 06:39:39 Skipping step because a previous step failed
2020/04/01 06:39:39 Skipping step because a previous step failed
```

**Comment by rhatdan** [comment]
This looks like the container is attempting to setup a user namespace now.  Are you running this container as non root?

**Comment by zhangtbj** [comment]
I tried both root and non-root, they all doesn't work....

I heard the buildah team plan to add non-privileged support?

Is there any news about it or is there any plan?

Thanks!

**Comment by rhatdan** [comment]
Depends on what you mean by non-privileged support?

Buildah needs to be able to create files owned by multiple UIDs  To do this we need to have "real" root or "user namespace" root.  

In order to to this the container needs to be either launched as root, in which case we can use Buildah with isolation --chroot in a locked down container.  IE One without --privileged flag.  This means the container gets the default Linux capabilities for a container.  These Capabilities were hard coded into Docker and containerd, they have a shorter list of Capabilities defined by CRI-O.  These capabilities include SETUID and SETGID, so that the buildah processes are able to create UID and GID files in the images. Note that this also requires a few additional Seccomp rules that are blocked by default by Docker/containerd, that we allow with CRI-O and Podman mainly `mount` and `unshare`.

The second option is to allow buildah to be run as nonroot but inside of a User Namespace.  This works similarly to running buildah directly as a non root user.  But in order to setup a user namespace to run buildah, it requires two File capabilties based tools newuidmap and newgidmap.  These tools require SETUID and SETGID to be able to preform their tasks.  So the container still needs to be launched with these two capabilities.  Again this type of container requires `mount` and `unshare` syscalls added to seccomp.json, if you are running with a Docker/containerd backend.

In the future, hopefully not two distant, we could use CRI-O to run the `buildah` container directly in a User Namespace, in which case it would already be running in a user namespace when it was launched and would require NO linux capabilities on the host, although it would still require Namespaced capabilities of `setuid` and `setgid` in order to support multiple uids defined inside of the user namespace.  We can do this now with Podman but it has not yet been implemented in Kubernetes.

**Comment by zhangtbj** [comment]
Thanks for your clear clarify! @rhatdan 

I am trying the first case:
- Use `runasroot` to the buildah
- enable the `SETUID` and `SETGID` capabilities
- Add `--isolation=chroot` for buildah bud command

But it still report error during bud:
```
Error during unshare(CLONE_NEWUSER): Operation not permitted
level=error msg="error parsing PID \"\": strconv.Atoi: parsing \"\": invalid syntax"
level=error msg="(unable to determine exit status)"
```

So I am wondering did I miss any other config?

**Comment by rhatdan** [comment]
It is still blocked by seccomp.json or docker/containerd that is running the back end on kubernetes.  I believe.  Even though you are using chroot, the buildah container still needs to `mount` the /proc and /sys file systems, this is only allowed when the container is running in a user namespace, which requires the unshare syscall, which is being blocked.

**Comment by rhatdan** [comment]
CRI-O and Podman allow unshare syscall by default.  Can you tell the pod to launch without seccomp filtering?

**Comment by zhangtbj** [comment]
Hi do you know how to launch without the seccomp filtering in kube PSP?

I tried to remove the these annotations from PSP:
```
    seccomp.security.alpha.kubernetes.io/allowedProfileNames: docker/default
    seccomp.security.alpha.kubernetes.io/defaultProfileName: docker/default
```

And tried to launch the Buildah build and got a new error:
```
Error: error writing "0 0 4294967295\n" to /proc/23/gid_map: write /proc/23/gid_map: operation not permitted
level=error msg="error writing \"0 0 4294967295\\n\" to /proc/23/gid_map: write /proc/23/gid_map: operation not permitted"
level=error msg="(unable to determine exit status)"
```

And I already provide the ["CHOWN", "DAC_OVERRIDE", "FOWNER"] capabilities.

**Comment by rhatdan** [comment]
The setting of GID_MAP and UID_MAP requires CAP_SETGID and CAP_SETUID respectively.
Sorry I don't know how to do this in Kubernetes.

**Comment by zhangtbj** [comment]
Hi Daniel,

I add the SETUID and SETGID capabilities:
```
        capabilities:
          add: ["CHOWN", "DAC_OVERRIDE", "FOWNER", "SETUID", "SETGID"]
```

And tried to run by using root (UID=0) and non-root (UID=1000), it changed to another new error....:
```
time="2020-04-06T01:55:50Z" level=warning msg="The cgroups manager is set to systemd but there is no systemd user session available"
time="2020-04-06T01:55:50Z" level=warning msg="For using systemd, you may need to login using an user session"
time="2020-04-06T01:55:50Z" level=warning msg="Alternatively, you can enable lingering with: `loginctl enable-linger 1000` (possibly as root)"
time="2020-04-06T01:55:50Z" level=warning msg="Falling back to --cgroup-manager=cgroupfs"
time="2020-04-06T01:55:50Z" level=warning msg="The cgroups manager is set to systemd but there is no systemd user session available"
time="2020-04-06T01:55:50Z" level=warning msg="For using systemd, you may need to login using an user session"
time="2020-04-06T01:55:50Z" level=warning msg="Alternatively, you can enable lingering with: `loginctl enable-linger 1000` (possibly as root)"
time="2020-04-06T01:55:50Z" level=warning msg="Falling back to --cgroup-manager=cgroupfs"
```

**Comment by rhatdan** [comment]
Other then these warnings, did it work?

**Comment by zhangtbj** [comment]
No it doesn't work.... :(

I forgot an error message in the end.

In the end, it report `permission denied`:
```
time="2020-04-06T01:55:50Z" level=warning msg="Falling back to --cgroup-manager=cgroupfs"
permission denied
```

**Comment by rhatdan** [comment]
https://github.com/containers/buildah/pull/2284
Should remove the warnings.

**Comment by rhatdan** [comment]
The Permission denied most likely is seccomp rules.

**Comment by rhatdan** [comment]
@umohnani8 @haircommander Any chance you could try to run buildahimage on cri-o to see if it works with Kubernetes?

**Comment by iwangxiaodong** [comment]
@rhatdan @zhangtbj 

k8s deployment add:
  spec.template.spec.containers[index].securityContext: **privileged: true**

**It works!**

**Comment by zhangtbj** [comment]
Hi @iwangxiaodong ,

We cannot use the `privileged: true` permission in our shared cluster, it is forbidden :(

**Comment by iwangxiaodong** [comment]
@zhangtbj 

```
docker run -it --rm --security-opt seccomp=unconfined \
    --security-opt apparmor=unconfined \
     quay.io/buildah/stable:latest bash
```

buildah pull busybox

It works!

k8s PodSecurityPolicy annotations  - https://kubernetes.io/docs/concepts/policy/pod-security-policy/#seccomp

**Comment by zhangtbj** [comment]
Thanks @iwangxiaodong ,

Let me try again.

**Comment by iwangxiaodong** [comment]
![image](https://user-images.githubusercontent.com/16535878/80634208-b120e900-8a8c-11ea-8039-a90a7419cde0.png)

@zhangtbj

We do not need：

    securityContext:
      privileged: true

**Comment by rhatdan** [comment]
LGTM
If you add a containers.conf to the image that specified to use cgroupfs rather then systemd, you can get rid of the warnings.

Something like:
https://github.com/containers/buildah/blob/master/contrib/buildahimage/stable/Dockerfile

**Comment by rhatdan** [comment]
I am going to close this since I believe you are satisfied.  Reopen if I am mistaken.

**Comment by tdensmore** [comment]
> @zhangtbj wrote:
> 
> ```
> docker run -it --rm --security-opt seccomp=unconfined \
>     --security-opt apparmor=unconfined \
>      quay.io/buildah/stable:latest bash
> ```
> 
> buildah pull busybox
> 
> It works!
> 
> k8s PodSecurityPolicy annotations - https://kubernetes.io/docs/concepts/policy/pod-security-policy/#seccomp

Unfortunately this (and every other suggestion) did not work for me:

```
buildah bud --isolation=chroot -v $PWD:/build/fuse-overlayfs -t fuse-overlayfs -f ./Containerfile.static.ubuntu .
STEP 1: FROM docker.io/ubuntu:rolling
Trying to pull docker.io/library/ubuntu:rolling...
Getting image source signatures
Copying blob 5c1f3479a03a done
Copying blob 74eb3aea4da4 done
Copying blob 2bfcfbcbcd59 done
Copying config 478aa0080b done
Writing manifest to image destination
Storing signatures
ERRO[0002] error unmounting /var/lib/containers/storage/overlay/c9a3bfc46dcfd3ce493bcbecd05f80f27252a39518b7f53a7c567f878e51e246/merged: invalid argument
error mounting new container: error mounting build container "48c19ead1403b9f2f2aaf328163618eef95d8c2082b0421969c3dc4cbbc4b6ae": error creating overlay mount to /var/lib/containers/storage/overlay/c9a3bfc46dcfd3ce493bcbecd05f80f27252a39518b7f53a7c567f878e51e246/merged, mount_data="nodev,fsync=0,lowerdir=/var/lib/containers/storage/overlay/l/LDSGZPQRPLGETZT7BHUJGYZ6EM:/var/lib/containers/storage/overlay/l/5XQYE63YFO27ZPOB34A3DV4P42:/var/lib/containers/storage/overlay/l/SQWFK7OAHITTHGTJBYLDSLTBFX,upperdir=/var/lib/containers/storage/overlay/c9a3bfc46dcfd3ce493bcbecd05f80f27252a39518b7f53a7c567f878e51e246/diff,workdir=/var/lib/containers/storage/overlay/c9a3bfc46dcfd3ce493bcbecd05f80f27252a39518b7f53a7c567f878e51e246/work,volatile": using mount program /usr/bin/fuse-overlayfs: fuse: device not found, try 'modprobe fuse' first
fuse-overlayfs: cannot mount: No such file or directory
: exit status 1
ERRO[0002] exit status 125
[root@a67f78bb329f fuse-overlayfs]#
```
